### ISSUE TITLE
Feature: Add 'Safe Mode' (ala LibreOffice) that offers comprehensive options to deal with user config files

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#6106]
Issue imported from https://tracker.freecad.org/view.php?id=4443

* **Reporter:** Kunda1
* **Date submitted:** 9/25/2020
* **FreeCAD version:** 
* **Category:** Feature
* **Status:** new
* **Tags:** #ready-for-migration, UI

# Original report text

Apparently LO has a cool feature that helps users deal with corrupt profile/config files. It's called **Safe Mode**. It backs up the current config files and uses a vanilla copy. This would be mean: No more asking users to back up their `~/.FreeCAD` and `~/.config/FreeCAD` directories (well at least the majority of usecases).

[Forum thread](https://forum.freecadweb.org/viewtopic.php?f=8&t=50342)

# Additional information

![image](https://user-images.githubusercontent.com/4140247/213080161-34747881-ec62-4854-ba2b-859441282051.png)

# Other bug information

* **Priority:** none
* **Severity:** feature
* **Category:** Feature
* **Updated:** 1/25/2022

# Discussion from Mantis ticket

### Comment by Kunda1 2021-02-13 15:05
Helpful findConfigFiles macro by TheMarkster that may be useful when creating this feature
https://forum.freecadweb.org/viewtopic.php?f=22&t=29888

---

### Comment by Kunda1 2021-03-09 00:16
https://github.com/furti/FreeCAD-PreferencesPresets

**Comment by adrianinsaval** [comment]
This could also use software OpenGL so that this can be easily tested

**Comment by luzpaz** [comment]
This feature code could also be utilized in #5772

**Comment by luzpaz** [comment]
This special mode can also be a way to start FreeCAD without 3rd party addons, so users and testers can help troubleshoot bugs whose usecase is related to an installed external addon.

**Comment by GAZ082** [comment]
This would be a nice feature.

**Comment by JohnOCFII** [comment]
Great idea.  OctoPrint, the 3D printing software tool also has a similar safe-mode feature.

**Comment by luzpaz** [comment]
Another usecase: 
* https://github.com/FreeCAD/FreeCAD/issues/16478#issuecomment-2346562289

**Comment by GAZ082** [comment]
> Another usecase:
> 
> * [Windows: Crash 'Access Violation' (1.0 rc1) #16478 (comment)](https://github.com/FreeCAD/FreeCAD/issues/16478#issuecomment-2346562289)

Thank you @luzpaz for making a case for this feature. We have a sizeable amount of issues triggered by "something" in the configs, which is really annoying.

**Comment by luzpaz** [comment]
> 20/20 hindsight I should have have somehow saved my broken config before resetting... 😕

Source: https://github.com/FreeCAD/FreeCAD/issues/16525#issuecomment-2351110671

**Comment by hyarion** [comment]
How about this:

* if freecad fails to start, it writes a special `FREECAD_SAFE_MODE` file to /tmp (or similar dir).
* on a successful load, it removes the special safe-mode file.
* if there's a safe-mode file available or if freecad is started with a new flag `--safe-mode`, then freecad loads into safe-mode.

Safe-mode could be the following:
* disable addons installed by user (skip the mod location that addon manager installs files into)
* avoid loading _and_ saving config files

Basically try to load freecad in a factory default state.

Would this be enough? Or would you like to change anything for a first iteration?

---

In a later iteration on this, we could extended it to load a special dialog to let the user preform different actions, similar to the one from the screenshot in the ticket.

**Comment by GAZ082** [comment]
@hyarion 
Hi Ben. When you refer to avoid loading a config file, you basically mean skipping all user modifications. I'm not very well versed on how configs work in FC, sorry if this seems to be a stupid question.

**Comment by luzpaz** [comment]
Hi @wwmayer, could you weigh in ? :pray: (https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2356931221)

**Comment by hyarion** [comment]
> When you refer to avoid loading a config file, you basically mean skipping all user modifications. 

Config file contains all settings, addons means plugins.
But the end result would be that we skip all user modifications so we get a clean state.

**Comment by luzpaz** [comment]
@Chennes do you have a moment to weigh-in on the proposal https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2356931221 ?

**Comment by chennes** [comment]
This seems fine to me. Can we also disable Qt's Settings store?

**Comment by luzpaz** [comment]
> Can we also disable Qt's Settings store?  

@hyarion ?

**Comment by hyarion** [comment]
Sure!

**Comment by luzpaz** [comment]
@hyarion can I assign you to this ticket ?

**Comment by hyarion** [comment]
@luzpaz feel free to assign it to me :)

**Comment by hyarion** [comment]
@chennes: `QSettings` seems a bit difficult to disable as it is used for a lot of different things.
For example, we still want it to be able to load ini files for us and get windows version from the windows registry.

We could let a new `FCSettings` subclass `QSettings` and let that handle if it should pass on the calls to super or just keeping the settings in memory during the session.
I do feel that this might be more complicated than what is worth.

Do you (or anyone else?) know why some code stores settings in `QSettings` while other code stores similar data as parameters in `user.cfg` using `App::GetApplication().GetParameterGroupByPath()` and similar?

Would it be an option to refactor the settings stored in QSettings to store the data in `user.cfg` instead?

**Comment by chennes** [comment]
`QSettings` is mostly used to store the stuff that Qt wants to store there (toolbar positions, window positions, etc.) while it would be delightful to rip it out entirely, that is a project in its own right. I'm happy with the new Safe Mode not touching it for the time being, it was a suggestions of a "nice to have" feature, not an essential component.

**Comment by hyarion** [comment]
I have a first implementation of the features discussed above

The implementation can be found in this branch: https://github.com/hyarion/FreeCAD/tree/safe-mode

Here is what it does:
* Creates a new temporary directory which is removed when the user exit freecad
* The following directories are mapped to new directories in the newly created temporary directory: `UserAppData`, `UserConfigPath`, `UserCachePath`, `AppTempPath`, `UserMacroPath`, `UserHomePath`

Safe mode can be entered in three ways:
* When `--safe-mode` is added as an argument when starting the program
* From menu -> help -> restart in safe mode
* If freecad crashed the last time (in the last x hours)

When start mode is enabled, it will appends ` (Safe Mode)` in the main window title and show a pop up after start informing the user that safe mode is entered.


What's missing is:
* QSettings (nice to have, can be added later)
* UI for configuration like in libreoffice, though I think we need to refine what we actually want before that can be implemented.
* Icon for command in help menu, reuses the `view-refresh` icon for now.

Caveats:
* The command in help is still enabled in safe mode, making it possible to restart into safe mode, from safe mode. (I haven't decided if this is a feature or bug)
* If the user preforms a change in safe mode that requires freecad to restart, it will restart in safe mode again, with the changes wiped (I'm not sure if this is a feature or bug either)

Screenshots:

Help menu:
<img width="489" alt="image" src="https://github.com/user-attachments/assets/2d46e41b-ef6b-4de7-acc6-f2e7af088bbc">

Popup after load ([qt on mac ignores the window title on dialogs, as required by apples guide lines](https://doc.qt.io/qt-6/qmessagebox.html#setWindowTitle)):
<img width="489" alt="image" src="https://github.com/user-attachments/assets/16f0f210-ca86-469e-bc8f-d99f6a30776d">

Window title:
<img width="254" alt="image" src="https://github.com/user-attachments/assets/007f8934-7142-403e-9d9c-fbd4ace74cd4">


What do you think, is this in line of what has been requested? Is there anything you want to be changed or added to the first iteration?
I will wait for feedback here before creating a merge request.

**Comment by luzpaz** [comment]
This is a huge first step! 
:wave: @pgilfernandez, do you have time to make a Safe Mode icon ? 
![Screenshot_20240922_075643](https://github.com/user-attachments/assets/fcdfcddc-f3a9-4036-bf6b-df0c8d272f41)

**Comment by GAZ082** [comment]
@luzpaz i could just change the color of the current icon to something like green real quick

**Comment by GAZ082** [comment]
![safe-mode](https://github.com/user-attachments/assets/7a2dbc93-9aa3-43c3-829b-405d66e10748)

**Comment by pgilfernandez** [comment]
Hi!,
Is enough with that color change or do you prefer to design a whole new one? I would do it if you need it

**Comment by hyarion** [comment]
> Is enough with that color change or do you prefer to design a whole new one? I would do it if you need it

I think the current one is good enough for the current version. But if you want to design one then I'll happily include it 😃

**Comment by luzpaz** [comment]
@GAZ082 IMHO, I think the choice of color, green, doesn't indicate to me of a 'safe mode' state.  Yellow or red seem more appropriate. 

@pgilfernandez maybe a new one ? :stuck_out_tongue_winking_eye:

**Comment by GAZ082** [comment]
@luzpaz i made it green because you know, safe? Red is bad! 😁

**Comment by JohnOCFII** [comment]
> @luzpaz i made it green because you know, safe? Red is bad! 😁
<img width="405" alt="Screenshot 2024-09-23 at 12 27 17 PM" src="https://github.com/user-attachments/assets/85d25267-54a2-43b1-87ba-bc940d4124f6">

maybe add a small green cross like the pharmacies in Europe

**Comment by GAZ082** [comment]
@JohnOCFII I thought about that until you realize that the cross is mostly a western thing in the Middle East you have the crescent instead.

**Comment by luzpaz** [comment]
![Screenshot_20240923_143041](https://github.com/user-attachments/assets/f992e49e-98aa-4e01-bdc2-e7bb0e5297fc)

**Comment by hyarion** [comment]
<img width="131" alt="image" src="https://github.com/user-attachments/assets/74bec9c5-c6a5-409c-b7cc-90ac99465ed9">

**Comment by adrianinsaval** [comment]
for icon choices please ask @FreeCAD/design-working-group

**Comment by hyarion** [comment]
@adrianinsaval how and where do I ask them?

**Comment by adrianinsaval** [comment]
by pinging the group as I did or on the discussion in the discord server https://discord.gg/w2cTKGzccC

**Comment by GAZ082** [comment]
asking as we speak

**Comment by obelisk79** [comment]
![image](https://github.com/user-attachments/assets/573c308c-22ca-4e1f-bee8-79e6c2299d65)

**Comment by GAZ082** [comment]
@obelisk79 not a fan of these as a cross may be widely recognized but it's not used in the middle east.

**Comment by MisterMakerNL** [comment]
After a quick search I find this icon a lot:
![safe-icon-16-2069139078](https://github.com/user-attachments/assets/3ec9bbe4-39c9-445f-b5cd-79df27a54b69)

**Comment by pgilfernandez** [comment]
> After a quick search I find this icon a lot:
> 
> ![safe-icon-16-2069139078](https://github.com/user-attachments/assets/3ec9bbe4-39c9-445f-b5cd-79df27a54b69)
> 
> 

Yes, this is the idea I was going to design, a green shield...

**Comment by hyarion** [comment]
> ![rescue2](https://private-user-images.githubusercontent.com/17993710/370108400-f3d6176f-b137-4e26-8ee2-e1841ef5e8e1.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjgzMzQsIm5iZiI6MTcyNzE2ODAzNCwicGF0aCI6Ii8xNzk5MzcxMC8zNzAxMDg0MDAtZjNkNjE3NmYtYjEzNy00ZTI2LThlZTItZTE4NDFlZjVlOGUxLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDA4NTM1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwZDZlYTFmYzhkOWJkMGFjOTEyODZhYWVkMDk5NjZiNGU0MDZhYTE5OWNkOTFjYjRiNzJjZWFjMDc5MDQ3ZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7Y_VUjLyxTN5JtVZO9ox6gIpcazg5jdjJgKCrZbR_Ow)

The red cross on a white background, ie. the reverse of the swiss flag, is protected by the geneva convention. Let's not violate it :)

**Comment by hyarion** [comment]
Thanks for all the input on icons, I really appreciate it!

I think many of them represent safe mode quite well, but in this case we primarily need an icon for the command/action to "restart in safe mode". Not for "Safe Mode" as a concept (though that could be added to the dialog after the restart).

Before deciding on an icon, please check the [comment I made with the screenshots](https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2366744455) that shows the context of where it will be used.

**Comment by GAZ082** [comment]
Easy peasy, lets do the green shield but replace the tick with the circled arrows of the restart icon.

**Comment by GAZ082** [comment]
Something like this:

![image](https://github.com/user-attachments/assets/10304206-7126-47af-9186-1437c4c54b82)

**Comment by tobiasfalk** [comment]
> Something like this:
> 
> ![image](https://github.com/user-attachments/assets/10304206-7126-47af-9186-1437c4c54b82)
> 

Could you make the arrows blue, like the original restart icon

**Comment by GAZ082** [comment]
That would be kinda ugly, but I can try. If you guys want I can do something tonight after kids go to bed and if people likes it, make a PR.

**Comment by GAZ082** [comment]
![image](https://github.com/user-attachments/assets/3cbdd001-7b8d-4535-bbba-2077fc26ac17)

Like?

**Comment by hyarion** [comment]
Scaling is handled automatically as freecad uses svg icons.

I'm not at designer so I might be the wrong person to give feedback here, but I'll try anyway ^^

I think the icon looks quite nice, but I think it doesn't quite follow [the artwork guidelines](https://wiki.freecad.org/Artwork_Guidelines), does it?

* The colors doesn't match with those in the green palette.
![image](https://github.com/user-attachments/assets/7302ccd7-bc37-4674-a1a9-6cc29c5da757)
* There's also a highlight missing right inside of the outline.
* I think a white arrows should outline should be the brightest white. Then the fill should probably be a gradient with the gray you to the outline color, similar to how you have the fill for the shield.

**Comment by GAZ082** [comment]
Followed the guidelines. Had to use the Tango pallete as a base because the FC one is not available (do not have access to the GDrive file).

I'll see if can do more changes today!

**Comment by hyarion** [comment]
@luzpaz https://github.com/FreeCAD/FreeCAD/issues/16834 another usecase

**Comment by luzpaz** [comment]
@hyarion maybe we should just merge to main as is and continue updating the functionality/esthetics as we go ?

**Comment by GAZ082** [comment]
I got access to the FC pallete, will have a look at it today and do a review. Then see if can add a PR to your brach @hyarion , what you say?

Just to get something started, then i can refine the icon (or someone else).

**Comment by hyarion** [comment]
Sure go for it!
If we don't manage to get the icon ready in time (not sure how icons are reviewed), we can always add the icon in the next iteration of the tool :)

**Comment by GAZ082** [comment]
@obelisk79 is there any formal icon review process?

**Comment by GAZ082** [comment]
@hyarion check this out 
![icon](https://github.com/user-attachments/assets/8329f243-3528-4021-988d-ca56048c2afb)

**Comment by hyarion** [comment]
Thanks for taking the time, I like the icon!

For the record, @GAZ082 popped in the discord channel so if anyone wants to help, feel free to pop by the design-working-group channel :)

**Comment by luzpaz** [comment]
> If we remove that feature, the only ways that safe-mode can get activated is manually through the help menu and by adding `--safe-mode` on the command line - and that should be safe to merge.

Yes, this is a smart compromise, imho

**Comment by GAZ082** [comment]
The guys at the discord toned down the gradient:

![image](https://github.com/user-attachments/assets/f4bad22e-7abc-4358-860d-7b5b2493e94e)

**Comment by hyarion** [comment]
Cool, great work!

**Comment by luzpaz** [comment]
Team work makes the dream work :fist_right: :fist_left:

**Comment by maxwxyz** [comment]
would it make sense to also implement a button to reset config files? E.g. on a fresh install it could also be asked in the start tab when FreeCAD detects a config file. This file becomes bloated easily it seems, and the safe mode only temporarily solves the issue, as there is no way to reset it?

**Comment by hyarion** [comment]
@maxwxyz 
> would it make sense to also implement a button to reset config files?

Short answer:
Yes

Long answer:
We just reduced the featurespace to create a MVP (the linked PR). By creating a MVP the idea is that we can get it out sooner to help people identify if their issue is the to their setup or a bug everyone has.

The feature you're asking for is important and a planned one, but the idea is to include that in the next iteration of safe mode. There we're also planning to bring back the feature to enter safe mode automatically if it detects a recent failed boot which was ripped out for the MVP (<--- this auto-safe-mode would have been needed for #16838)

I'm leaning towards keeping the linked PR as is, but I would also like to start the discussion on what the next step is (this is where I think your request comes in).

I think that libreoffice' safe mode dialog has a lot of nice functionality but the dialog is a bit too intimidating to look at with all these options. I would like if we could avoid creating a carbon copy of it, but rather trying to identify the most important features that we would like to include.

---

Roadmap: Safe-mode 2.0:
* automatically enter safe mode when finding a recent boot-crash. <-- help users who can't even open freecad like #16838
* easy way for user to do a factory reset: remove all configuration files + addons (+ more?) <-- this sledge hammer would help everyone that has a broken setup
* (what else should we include here?)

**Comment by maxwxyz** [comment]
great. Could you create new issues for the followup?

**Comment by hyarion** [comment]
The title of this issue is:
> Feature: Add 'Safe Mode' (ala LibreOffice) that offers comprehensive options to deal with user config files

I can create followup issues if you want @maxwxyz, but personally I think we can keep working on this one as it would be a copy of the title and description anyway.

Let me know if you want me to create a new one anyway.

**Comment by luzpaz** [comment]
@hyarion i'll create the issue if necessary. Although, I'm in agreement that we keep working on it as we go without creating issues for now.

**Comment by luzpaz** [comment]
Does anyone have a proposal of what a UI dialog would look like for a FreeCAD Safe Mode ?

**Comment by hyarion** [comment]
> Does anyone have a proposal of what a UI dialog would look like for a FreeCAD Safe Mode ?

I heard someone called Kunda1 on the forum had done some work, maybe even created a qt .ui file? ;)

---

But before designing an UI, do we have any more requirements on the functionality? (5 posts up, search for "Roadmap: Safe-mode 2.0")

**Comment by luzpaz** [comment]
> I heard someone called Kunda1 on the forum had done some work, maybe even created a qt .ui file? ;)

Indeed. There wan an attempt, lol

> But before designing an UI, do we have any more requirements on the functionality? (5 posts up, search for "Roadmap: Safe-mode 2.0")

Yes, you're right. I don't mean to jump the gun. 

>> Roadmap: Safe-mode 2.0:
>> ...
>> (what else should we include here?)

* Restore from backup 
* Reset factory settings (deletes configs)
* Disable any hardware related options (hardware acceleration etc...)
* Backup current configs
* Manually edit configs (expert mode)
  * Perhaps have a link or supplemental UI for [Fine-tuning](https://wiki.freecad.org/Fine-tuning) related tweaks
* Addons
   * Disable all addons
   * Disable specific addons via secondary UI listing all installed addons  
   * Delete specific addons
* Conveniently displaying [About](https://wiki.freecad.org/About) dialog info (collapsed by default)

**Comment by hyarion** [comment]
@luzpaz this got closed because the MVP was merged today. Should we reopen this one or create a new one?

**Comment by luzpaz** [comment]
What's your preference ?

**Comment by hyarion** [comment]
Reopen this, but I have no strong opinion on it

**Comment by luzpaz** [comment]
Active entries:  
* https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2378722953
* https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2423342525

**Comment by luzpaz** [comment]
Relevant comment: https://github.com/FreeCAD/FreeCAD/issues/6106#issuecomment-2423342525

**Comment by leoheck** [comment]
"Continue in safe mode" seems convoluted. 

It could be a simple cancel button.
